Issue 2378 lang level feed config

SumDonkuS opened the pull request:
Sanity check:

* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/getzola/zola/pulls) for the same update/change?

## Code changes

Looking at [issue 2378](https://github.com/getzola/zola/issues/2378) and the related [issue 2360](https://github.com/getzola/zola/issues/2360), found that there were a few places that needed to check the LangaugeOptions.  To handle the change for detecting the settings for `feed_filename`, the code has been switched to having this config be optional.  This also allowed for switching back to the first implementation of the `merge_field` macro. `generate_feed` checks both the section for the default language as well as the global area for its config.  If it exists in both places an error is thrown for the collision.

* [x] Are you doing the PR on the `next` branch?

* [x] Have you created/updated the relevant documentation page(s)?

No documentation change for this bug fix.

This is my first PR for this project.  Please let me know if there is anything that I'm missing.  Thank you again for the great project.

SumDonkuS committed:
initial code changes to check language options for feed_filename and generate feed for default language

SumDonkuS committed:
revert unneeded modifications for handling the generate feed logic for default and other languages.

Keats commented on components/config/src/config/mod.rs:
```diff
@@ -202,12 +203,18 @@ impl Config {
+                .feed_filename
+                .as_ref()
+                .map(|ff| path.ends_with(ff))
+                .unwrap_or_else(|| path.ends_with(DEFAULT_FEED_FILENAME))
```
I would maybe put that as a variable above so the condition is easier to read.

SumDonkuS replied:
Good call.  I have that one changed to keep the if statement more readable.

Keats commented on components/site/src/lib.rs:
```diff
@@ -749,7 +749,13 @@ impl Site {
+                .config
+                .languages
+                .get(&self.config.default_language)
+                .map_or(false, |lang_opt| lang_opt.generate_feed)
```
looks like we do need better config querying based on a language like you did with `language_feed_filename`.
No need to change anything here, just thinking aloud

SumDonkuS replied:
Let me know if the change I have is what you were thinking.

SumDonkuS commented:
Thank you for catching those things for me.  I'll keep a closer eye out for those in the future.  Let me know if the modification to querying `generate_feed` from the config looks like what you were thinking.  Thank you again.

SumDonkuS commented on components/site/src/lib.rs:
```diff
@@ -1022,6 +1029,19 @@ impl Site {
         Ok(())
     }
 
+    pub fn language_feed_filename(&self, lang: &str) -> String {
```
Good call.  Completely spaced that.  Thank you for the heads up.

Keats commented on components/config/src/config/languages.rs:
```diff
@@ -18,7 +18,7 @@ pub struct LanguageOptions {
     pub generate_feed: bool,
     /// The filename to use for feeds. Used to find the template, too.
     /// Defaults to "atom.xml", with "rss.xml" also having a template provided out of the box.
```
need to change the docstring no?

Keats commented on components/config/src/config/mod.rs:
```diff
@@ -66,7 +67,7 @@ pub struct Config {
     pub feed_limit: Option<usize>,
     /// The filename to use for feeds. Used to find the template, too.
     /// Defaults to "atom.xml", with "rss.xml" also having a template provided out of the box.
```
same here

Keats reviewed (commented):
I think it's ok. A bit iffy with Option<String> but maybe it's fine

SumDonkuS commented:
Thank you for spotting those docs.  They should now be modified to describe how the default values are handled.

Keats commented:
Note: that PR conflicts with https://github.com/getzola/zola/pull/2477, i'll have a look at rebasing it after 2477 is merged

Keats commented:
There seems to be more spots to fix than I thought. I'll release the next version of Zola first and get back to that after, otherwise I'm never going to do that release.

Keats committed:
Next version

SumDonkuS commented:
Sorry for the delay.  Work pulled me away for a bit.

kytta committed:
Fix YAML date parsing (#2549)

* Refine YAML date regex

This commit does a few changes:

- Introduce a new regex
  - it is a bit off-spec (it allows one-digit months and days in date-only mode)
  - uses named groups
  - avoids group duplication
- parses offset once

Fixes #2538

* Fix nanosecond parsing

* Rename variables for brewity

* Add tests

Keats committed:
Fix some feed_filenames issues with multi languages

Keats commented:
I've put a tiny commit on the next branch: https://github.com/getzola/zola/pull/2547/commits/4c5cc5399b1b0f6590b6b7e964166107ed7298dd for some obvious issues.
Do you want to restart work on that PR? Probably better to start from scratch tbh

Keats committed:
Ignore Helix temp files

SumDonkuS committed:
Remove debugging print.  Add test for generate_feeds defined in both global and default language block.  Currently set to prefer the language level config.

SumDonkuS committed:
Fix comment.  Remove debug println.  Correct modification to serialized config.

SumDonkuS committed:
Fix comment

SumDonkuS committed:
formatting fixes

SumDonkuS commented:
Sorry.  Missed your comment about starting fresh after rebasing.  Let me know if you would like to start from scratch.

Keats commented on components/site/src/lib.rs:
```diff
@@ -744,27 +744,7 @@ impl Site {
-            self.render_feeds(pages, Some(&PathBuf::from(code)), code, |c| c)?;
-            start = log_time(start, "Generated feed in other language");
-        }
+        self.render_all_feeds()?;
```
This name is not accurate since there are also section feeds. `render_site_feeds`?

Keats commented on components/site/src/lib.rs:
```diff
@@ -784,6 +764,41 @@ impl Site {
+            let is_default_language = code == &self.config.default_language;
+
+            let skip_default_language_feed_generation =
+                is_default_language && !self.config.generate_feeds && !language.generate_feeds;
```
should that ambiguity be resolved when loading the config? Eg error if the value at the top level and in a default language level differs.
This way you can now just do something like `self.config.languages.iter().filter(|x| x.generate_feeds)` for the loop

Keats commented on components/site/src/lib.rs:
```diff
@@ -1209,3 +1226,200 @@ fn log_time(start: Instant, message: &str) -> Instant {
 }
+
+#[cfg(test)]
+mod tests {
```
Can those be moved in the tests folders? This file is already long enough without tests
